dracut/30ignition/persist-ifcfg.sh

```diff
@@ -33,6 +33,16 @@ persist_ifcfg() {
         return 0
     fi
 
+    # Only persist the hostname if something else has not populated
+    # /etc/hostname, such as Ignition Files.
+    hname="$(< /proc/sys/kernel/hostname)"
+    hpath="/sysroot/etc/hostname"
+    if [ ! -e "${hpath}" ] && [ -n "${hname}" ]; then
```

Comment:
Ah, good catch yes, we don't want to set the persistent hostname if it's localhost.

The whole "which interface provided a hostname" is actually a mess, I bet in NetworkManager today it's just "first one wins".

(The topic of static setups but still having DHCP enabled on lingering other devices is touched on in [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=1800900))

Reply:
I'll update it to catch that case. The prior version used a check on that but that was dropped because DHCP persistent was deemed undesirable. 
```
ConditionHost=!localhost
ConditionHost=!localhost.localdomain
```

Reply:
Most elegant I think would be if dracut wrote `/etc/hostname` in the initrd only for persistent cases, then we could just mirror that.  I don't know why it isn't; probably just an oversight.

Reply:
I suspect that the reason why dracut does not write `/etc/hostname` is because it doesn't know where the hostname comes from: it uses a udev rule to configure the device when it appears. The last device to come is used to set the hostname. The deeper I look at it, the more I wonder why there isn't a `hostname=` karg.

Reply:
> One thing to note is that systemd will set a default transient hostname at startup, so `[ -n "${hname}" ]` will never actually fail.
> 
> IOW, this changes the behaviour subtly in the case where you set a static IP via `ip=` but _not_ a hostname: before, we'd boot all the way without any `/etc/hostname`, now we'll boot with it set to e.g. `localhost`.

Actually, no. I tested that use-case. The parts of systemd that would do that is not present in the initramfs at this point. You need `systemd-hostnamed` which is a socket listener. I evaluated that path in the last commit. Further, in testing, I confirmed that `/etc/hostname` was missing booting without an `ip=` provided hostname.

Reply:
> Actually, no. I tested that use-case. The parts of systemd that would do that is not present in the initramfs at this point.

Hmm, weird. It's basically the very first thing systemd does on both FCOS and RHCOS:

```
Welcome to Fedora CoreOS 31.20200306.dev.5 dracut-049-27.git20181204.fc31.1 (Initramfs)!

[    2.142750] systemd[1]: No hostname configured.
[    2.144456] systemd[1]: Set hostname to <localhost>.
```

And checking in `rd.break`:

```
switch_root:/# cat /proc/sys/kernel/hostname
localhost
```